### PR TITLE
Language change

### PR DESCRIPTION
**Pull request by Sarahhud**
Added a dropdown menu of 30 languages that the user can select from to change the language of the historical text.

### CODE REVIEW

File: capstone-app/src/app/modules/history/WikiServiceResult.ts [hunk 1]
```diff
@@ -2,4 +2,5 @@ export interface WikiServiceResult {
   title: string,
   history: string,
   furtherReading?: Map<string, string>
+  langlinks?: Map<string, Map<string, string>>
```
Comment (billyporter) [inline capstone-app/src/app/modules/history/WikiServiceResult.ts:5]: It's a little confusing what should go in this map of maps, wouldn't an interface be more appropriate for readability?

File: capstone-app/src/app/modules/history/history.module.ts [hunk 2]
```diff
@@ -1,14 +1,19 @@
...
 import { WikiMediaComponent } from './wiki-media/wiki-media.component';
-import { FormsModule } from '@angular/forms';
 import { HttpClientModule } from '@angular/common/http';
 import { MatCardModule } from '@angular/material/card';
 import { MaterialModule } from './../material/material.module';
+import {FormsModule, ReactiveFormsModule} from '@angular/forms';
+import {MatNativeDateModule} from '@angular/material/core';
+import {BrowserModule} from '@angular/platform-browser';
+import { MatMenuModule } from '@angular/material/menu';
+import {MatIconModule} from '@angular/material/icon';
```
Comment (billyporter) [inline capstone-app/src/app/modules/history/history.module.ts]: These are already included in the Material Module Imports. Please move these imports to the material module and then import in the material module. I believe I mentioned this in your last PR.

File: capstone-app/src/app/modules/history/wiki-media/wiki-media.component.html [hunk 3]
```diff
@@ -1,3 +1,4 @@
+
 <head>
   <link href="https://fonts.googleapis.com/css2?family=Roboto&display=swap" rel="stylesheet">
```
Comment (billyporter) [inline capstone-app/src/app/modules/history/wiki-media/wiki-media.component.html]: Since we're now using Roboto across the entire app, it may be more appropriate to place this in index.html

File: capstone-app/src/app/modules/history/wiki-media/wiki-media.component.html [hunk 4]
```diff
@@ -15,7 +16,26 @@
...
   {{error}}
+  <button mat-icon-button [matMenuTriggerFor]="menu">
+    <mat-icon>language</mat-icon>
+    <span>Languages</span>
+  </button>
+  <mat-menu #menu="matMenu" class="languages">
+    <button *ngFor="let lang of languages | keyvalue" mat-menu-item (click)="onChangeLanguage(lang.key)">
+      {{lang.key}}
+    </button>
+  </mat-menu>
```
Comment (billyporter) [inline capstone-app/src/app/modules/history/wiki-media/wiki-media.component.html]: Is there a reason we have this included twice? (lines 19-27)

File: capstone-app/src/app/modules/history/wiki-media/wiki-media.component.spec.ts [hunk 5]
```diff
@@ -53,12 +125,39 @@ describe('WikiMediaComponent', () => {
...
+
+    it('title is changed correctly on changeLanguage call', () => {
+      component.changeLanguage('Stillwater, (Oklahoma)', 'de', 'Geschichte');
+      expect(component.title).toEqual(testWikiGermanServiceResponse.title);
+    });
+
+    it('langlinks are changed correctly on changeLanguage call', () => {
+      component.changeLanguage('Stillwater, (Oklahoma)', 'de', 'Geschichte');
+      expect(component.langlinks).toEqual(testWikiGermanServiceResponse.langlinks);
+    });
```
Comment (billyporter) [inline capstone-app/src/app/modules/history/wiki-media/wiki-media.component.spec.ts]: In addition to checking if the langlinks variable has the correct content, it may be a good idea to include another test checking if the mat menu items on the dom update and match the new content of the langlinks

File: capstone-app/src/app/modules/history/wiki-media/wiki-media.component.ts [hunk 6]
```diff
@@ -17,6 +21,15 @@ export class WikiMediaComponent implements OnChanges, OnInit {
   loading = true;
   body: string;
   urls = new Map();
+  langlinks = new Map();
+  languages: any = Languages as any;
+  language: any = Languages as any;
```
Comment (billyporter) [inline capstone-app/src/app/modules/history/wiki-media/wiki-media.component.ts]: Wouldn't this set both language and languages to the same object? Additionally, can we get more specific than " as any"?

File: capstone-app/src/app/modules/history/wiki-media/wiki-media.component.html [hunk 7]
```diff
@@ -15,7 +16,26 @@
       <a href="{{url.key}}" target="_blank">{{url.value}}</a>
     </p>
   </mat-card>
+  <button mat-icon-button [matMenuTriggerFor]="menu">
+    <mat-icon>language</mat-icon>
+    <span>Languages</span>
+  </button>
+  <mat-menu #menu="matMenu" class="languages">
+    <button *ngFor="let lang of languages | keyvalue" mat-menu-item (click)="onChangeLanguage(lang.key)">
```
Comment (billyporter) [inline capstone-app/src/app/modules/history/wiki-media/wiki-media.component.html]: Wouldn't this set the list of buttons to "en, es, fr,... " etc. Can you use a map to have the names be "English, Spanish, French.." etc?

File: capstone-app/src/app/modules/history/wiki-media/wiki-media.component.spec.ts [hunk 8]
```diff
@@ -53,12 +125,39 @@ describe('WikiMediaComponent', () => {
 
     it('displays location history', () => {
       component.getResults('Stillwater, Oklahoma');
-      expect(component.history).toEqual(testWikiResult.parse.text["*"]);
+      expect(component.history).toEqual(testWikiServiceResponse.history);
     });
 
     it('displays correct location title', () => {
```
Comment (billyporter) [inline capstone-app/src/app/modules/history/wiki-media/wiki-media.component.spec.ts:129]: For this test and the "displays location history one" it makes more sense to either 1) switch the name of the test or 2) look at the DOM to see if the info actually is displayed

File: capstone-app/src/app/modules/history/wiki-media/wiki-media.component.ts [hunk 9]
```diff
@@ -34,19 +47,69 @@ export class WikiMediaComponent implements OnChanges, OnInit {
...
+  onChangeLanguage(prefix: string) {
+    if (prefix == this.prevPreFix) {
+      this.changeLanguage(this.prevQuery, this.prevPreFix, this.prevWordForHistory);
+    }
+    else if (prefix == 'en' && !(this.langlinks.get('en'))) {
+      this.prevPreFix = prefix;
+      const language = this.langlinks.get(prefix);
+      this.query = language.get('searchQuery');
+      this.prevQuery = this.query;
+      this.language = 'English';
```
Comment (billyporter) [inline capstone-app/src/app/modules/history/wiki-media/wiki-media.component.ts]: It's a little confusing to have a local variable language and a class variable language. Additionally, didn't you set the class variable language to equal the Languages object and now it's being assigned a string?

File: capstone-app/src/app/modules/history/wiki-media/wiki-media.component.ts [hunk 10]
```diff
@@ -34,19 +47,69 @@ export class WikiMediaComponent implements OnChanges, OnInit {
...
+      this.prevWordForHistory = 'History';
+      this.changeLanguage(this.query, 'en', 'History');
+    }
+    else {
+      const language = this.langlinks.get(prefix);
+      this.prevPreFix = prefix;
+      this.query = language.get('searchQuery');
+      this.prevQuery = this.query;
+      this.language = language.get('langName');
```
Comment (billyporter) [inline capstone-app/src/app/modules/history/wiki-media/wiki-media.component.ts]: As reference to my previous comment, it seems a little weird to have this.language be a json object then assigning it to a string.

File: capstone-app/src/app/modules/history/wiki-media/wiki-media.component.ts [hunk 11]
```diff
@@ -34,19 +47,69 @@ export class WikiMediaComponent implements OnChanges, OnInit {
...
     });
   }
+
+  changeLanguage(queryString: string, languagePrefix: string, wordForHistory: string) {
+    this.wikiService.searchNewLang(queryString, languagePrefix, wordForHistory)
+      .subscribe((result: WikiServiceResult) => {
+        this.history = result.history;
+        if (!(this.history) || (this.history.includes("CPU time usage:"))){
+          this.loading = true;
+          this.error = "Unfortunately, the language you requested does not have an available tranlation for this page. Please select a different language.";
```
Comment (billyporter) [inline capstone-app/src/app/modules/history/wiki-media/wiki-media.component.ts]: I'm a little worried about this object. I'd assume that for many of the languages listed, there may not be a "history" section, but there still would be informaiton, such as https://mg.wikipedia.org/wiki/Los_Angeles. Wouldn't it make more sense to include the intro paragraphs if there's no history section found?

Comment (billyporter) [reply capstone-app/src/app/modules/history/wiki-media/wiki-media.component.ts]: Logic* not object

File: capstone-app/src/app/modules/history/wiki-results.service.spec.ts [hunk 12]
```diff
@@ -50,8 +51,20 @@ describe('WikiResultsService', () => {
     req.flush(testResponse);
   });
 
+  it('searchNewLang(query) should return data', () => {
+    service.searchNewLang('Stillwater, (Oklahoma)', 'de', 'Geschichte').subscribe((result) => {
+      expect(result).toBeTruthy();
+    });
```
Comment (billyporter) [inline capstone-app/src/app/modules/history/wiki-results.service.spec.ts:56]: Can we have this be more specific? Right now it's passing if any data at all is returned. Would it make more sense to make sure the response is sensible?

File: capstone-app/src/app/modules/history/wiki-results.service.ts [hunk 13]
```diff
@@ -45,46 +53,72 @@ export class WikiResultsService {
...
+    return this.http.get<WikiSearchResult>(
+            `${WikiResultsService.URL_BEGINNING}${language}${WikiResultsService.URL_MIDDLE}${query}${WikiResultsService.URL_END}`
+          ).pipe(map(history => {
+          if (history.parse.text) {
+            const langlinks = new Map<string, Map<string, string>>();
+            for (let link of Object.values(history.parse.langlinks)) {
+                let links = new Map<string, string>();
+                links.set("langName", link.langname);
+                links.set("url", link.url);
+                links.set("searchQuery", link["*"]);
```
Comment (billyporter) [inline capstone-app/src/app/modules/history/wiki-results.service.ts]: single quotoes

File: capstone-app/src/app/modules/history/wiki-results.service.ts [hunk 14]
```diff
@@ -45,46 +53,72 @@ export class WikiResultsService {
...
-  fixString(text: string): {history: string, furtherReading?: Map<string, string>} {
-    const firstIndex = text.indexOf('<span class="mw-headline" id="History">History</span>', 0);
+  searchNewLang(query: string, language: string, wordForHistory: string): Observable<WikiServiceResult> {
+    return this.http.get<WikiSearchResult>(
+            `${WikiResultsService.URL_BEGINNING}${language}${WikiResultsService.URL_MIDDLE}${query}${WikiResultsService.URL_END}`
+          ).pipe(map(history => {
+          if (history.parse.text) {
+            const langlinks = new Map<string, Map<string, string>>();
+            for (let link of Object.values(history.parse.langlinks)) {
+                let links = new Map<string, string>();
```
Comment (billyporter) [inline capstone-app/src/app/modules/history/wiki-results.service.ts]: const

File: capstone-app/src/app/modules/history/wiki-results.service.ts [hunk 15]
```diff
@@ -45,46 +53,72 @@ export class WikiResultsService {
...
 
-  fixString(text: string): {history: string, furtherReading?: Map<string, string>} {
-    const firstIndex = text.indexOf('<span class="mw-headline" id="History">History</span>', 0);
+  searchNewLang(query: string, language: string, wordForHistory: string): Observable<WikiServiceResult> {
+    return this.http.get<WikiSearchResult>(
+            `${WikiResultsService.URL_BEGINNING}${language}${WikiResultsService.URL_MIDDLE}${query}${WikiResultsService.URL_END}`
+          ).pipe(map(history => {
+          if (history.parse.text) {
+            const langlinks = new Map<string, Map<string, string>>();
+            for (let link of Object.values(history.parse.langlinks)) {
```
Comment (billyporter) [inline capstone-app/src/app/modules/history/wiki-results.service.ts]: const

**Review by billyporter — CHANGES_REQUESTED** [review]
Ultimately looks really great! The one thing I am worried about is that this PR makes a lot of modifications to the service file and wiki component ts file. Since this PR has started, we've merged 5 PRs to master, a few of which involved the wiki component and tests. Before you merge this, I'd highly recommend that you create a copy, pull the changes from master, and makre sure the app still functions fine and that tests aren't failing.

(Especially see how we have a few conflicts with master branch)

File: capstone-app/src/app/modules/history/wiki-results.service.spec.ts (hunk 12 above)
Comment (Sarahhud) [reply capstone-app/src/app/modules/history/wiki-results.service.spec.ts:56]: Not really, it's doing very literally the exact same thing as the main search function (which is already tested) except using your wikiSearchHandler. I just want to make sure it makes a valid API call. The formatting of the string is already being tested.

Comment (Sarahhud) [reply capstone-app/src/app/modules/history/wiki-results.service.spec.ts:56]: Since only fake data would be returned, adding a more specific test wouldn't accomplish much in my opinion.

File: capstone-app/src/app/modules/history/wiki-media/wiki-media.component.ts (hunk 11 above)
Comment (Sarahhud) [reply capstone-app/src/app/modules/history/wiki-media/wiki-media.component.ts]: Yes, this actually would still do that. The var name is a little confusing, but the history string is returned with either the history section, or if that is null, history is returned with the beginning paragraphs of the wiki page. The only reason history would be null is if there was no text response on the wiki pg, or it loaded but didn't have any data in which case it returns cpu stats.

File: capstone-app/src/app/modules/history/wiki-media/wiki-media.component.spec.ts (hunk 5 above)
Comment (Sarahhud) [reply capstone-app/src/app/modules/history/wiki-media/wiki-media.component.spec.ts]: To be honest, I'm not sure how to do that. I would need someone to hop on a call with me and walk through this, but would be fine  doing that if other people think this needs to be added.

File: capstone-app/src/app/modules/history/wiki-media/wiki-media.component.ts (hunk 11 above)
Comment (billyporter) [reply capstone-app/src/app/modules/history/wiki-media/wiki-media.component.ts]: Can you go to the 'mg' translation of Los Angeles on your branch? The wikipedia article has information but it's displaying 'The Malagasy translation for Los Angeles is unavailable.
Unfortunately, the language you requested does not have an available tranlation for this page. Please select a different language' on our website

File: capstone-app/src/app/modules/history/wiki-media/wiki-media.component.html (hunk 4 above)
Comment (Sarahhud) [reply capstone-app/src/app/modules/history/wiki-media/wiki-media.component.html]: Yes, one only loads if history loads. If history isn't valid and there is nothing to show I still want the language change to be possible, bc the lack of data might actually be bc there is just no valid translation for that language. The formatting was better for the drop-down, I didn't spend a ton of time on it bc I knew you were changing the layout of stuff. All of that to say that one can definitely be removed as long as it's outside of the ngif statement.

File: capstone-app/src/app/modules/history/wiki-results.service.spec.ts (hunk 12 above)
Comment (billyporter) [reply capstone-app/src/app/modules/history/wiki-results.service.spec.ts:56]: There's a lot of logic in the searchNewLang function and there's no test to see if the logic behaves correctly. Couldn't we use this test with this fake data to make sure that the fake data is parsed how we would expect?

File: capstone-app/src/app/modules/history/wiki-media/wiki-media.component.html (hunk 4 above)
Comment (billyporter) [reply capstone-app/src/app/modules/history/wiki-media/wiki-media.component.html]: Thanks. Makes sense, all good here.

File: capstone-app/src/app/modules/history/WikiServiceResult.ts (hunk 1 above)
Comment (Sarahhud) [reply capstone-app/src/app/modules/history/WikiServiceResult.ts:5]: Maybe, but at this point it would be much simpler to just leave it like that.

File: capstone-app/src/app/modules/history/wiki-media/wiki-media.component.ts (hunk 11 above)
Comment (Sarahhud) [reply capstone-app/src/app/modules/history/wiki-media/wiki-media.component.ts]: Hmm, yes I'll look into that. It might be that sometimes there is data and a cpu statistic is returned?

File: capstone-app/src/app/modules/history/wiki-results.service.spec.ts (hunk 12 above)
Comment (Sarahhud) [reply capstone-app/src/app/modules/history/wiki-results.service.spec.ts:56]: What I mean is that the functions that are called to parse that data are the exact same ones that are already being testes to ensure they work. We could add another test thought, I just wanted to explain why I didn't think it was necessary.

File: capstone-app/src/app/modules/history/wiki-media/wiki-media.component.spec.ts [hunk 16]
```diff
@@ -1,38 +1,110 @@
...
-            + 'glorious prairie spreading out beneath the golden beams of an autumnal sun. The deep and frequent '
-            + 'traces of buffalo, showed it to be a one of their favorite grazing grounds.”9'
+    testWikiServiceResponse = {
+            title: 'Stillwater, Oklahoma',
+            history: 'The north-central region of Oklahoma became part of the United States with the Louisiana '
+              + 'Purchase in 1803. In 1832, author and traveler Washington Irving provided the first recorded '
+              + 'description of the area around Stillwater in his book A Tour on the Prairies. He wrote of “a '
+              + 'glorious prairie spreading out beneath the golden beams of an autumnal sun. The deep and frequent '
+              + 'traces of buffalo, showed it to be a one of their favorite grazing grounds.”9',
+           langlinks: new Map()
```
Comment (smontana30) [inline capstone-app/src/app/modules/history/wiki-media/wiki-media.component.spec.ts]: fix spacing for langlinks because it's not lined up with title and history

File: capstone-app/src/app/modules/history/wiki-media/wiki-media.component.spec.ts (hunk 5 above)
Comment (Sarahhud) [reply capstone-app/src/app/modules/history/wiki-media/wiki-media.component.spec.ts]: Billy and I talked about this and bc of the time crunch we are on and the fact that I can visually test and see the languages updating we won't be adding a test at the moment for this. If I have time later I will come back and add one.

File: capstone-app/src/app/modules/history/history.module.ts [hunk 17]
```diff
@@ -1,14 +1,15 @@
 import { NgModule } from '@angular/core';
 import { CommonModule } from '@angular/common';
 import { WikiMediaComponent } from './wiki-media/wiki-media.component';
-import { FormsModule } from '@angular/forms';
 import { HttpClientModule } from '@angular/common/http';
-import { MatCardModule } from '@angular/material/card';
 import { MaterialModule } from './../material/material.module';
+import { FormsModule, ReactiveFormsModule } from '@angular/forms';
+import { BrowserModule } from '@angular/platform-browser';
+
```
Comment (mbehrlich) [inline capstone-app/src/app/modules/history/history.module.ts:8]: Nit: Remove extra line of whitespace.

File: capstone-app/src/app/modules/history/wiki-media/wiki-media.component.spec.ts [hunk 18]
```diff
@@ -1,49 +1,121 @@
...
+              + 'description of the area around Stillwater in his book A Tour on the Prairies. He wrote of “a '
+              + 'glorious prairie spreading out beneath the golden beams of an autumnal sun. The deep and frequent '
+              + 'traces of buffalo, showed it to be a one of their favorite grazing grounds.”9',
+            langlinks: new Map()
         }
-      }
+
+        const englishLangLinks = new Map<string, Map<string, string>>();
+        const enDe = new Map<string, string>();
+        enDe.set("searchQuery", "Stillwater, (Oklahoma)");
```
Comment (mbehrlich) [inline capstone-app/src/app/modules/history/wiki-media/wiki-media.component.spec.ts]: Use single quotes in TypeScript, here and in all of the lines below.

File: capstone-app/src/app/modules/history/wiki-media/wiki-media.component.spec.ts [hunk 19]
```diff
@@ -1,49 +1,121 @@
...
+      const link = MockWikiServiceGermanResponse.langlinks[lang];
+      const links = new Map<string, string>();
+      links.set('searchQuery', link.searchQuery);
+      links.set('langName', link.langName);
+      links.set('url', link.url);
+      mockGermanResponse.langlinks.set(lang, links);
+    }
+
+    const wikiService = jasmine.createSpyObj('WikiResultsService', ['search', 'searchNewLang']);
+    getWikiResponseSpy = wikiService.search.and.returnValue(of(mockResponse));
```
Comment (mbehrlich) [inline capstone-app/src/app/modules/history/wiki-media/wiki-media.component.spec.ts]: Since getWikiResponseSpy and getWikiLangResponseSpy aren't referenced anywhere, they're not necessary. You can delete where they're defined above and these lines can just be:
```
wikiService.search.and.returnValue(of(mockResponse));
wikiService.searchNewLang.and.returnValue(of(mockGermanResponse));
```

File: capstone-app/src/app/modules/history/wiki-media/wiki-media.component.spec.ts [hunk 20]
```diff
@@ -53,12 +125,38 @@ describe('WikiMediaComponent', () => {
...
+      expect(component.langlinks).toEqual(testWikiServiceResponse.langlinks);
+    });
+
+    it('language change function is called on button selection', () => {
+      spyOn(component, 'changeLanguage');
+      component.onChangeLanguage('de');
+      expect(component.changeLanguage).toHaveBeenCalledWith('Stillwater, (Oklahoma)', 'de', 'Geschichte');
+    });
+
+    it('history is changed correctly on changeLanguage call', () => {
```
Comment (mbehrlich) [inline capstone-app/src/app/modules/history/wiki-media/wiki-media.component.spec.ts:145]: Optional: This is controversial, but since these three tests have identical setup, you can have just one test with three expects:

```
    it('component is changed correctly on changeLanguage call', () => {
      component.changeLanguage('Stillwater, (Oklahoma)', 'de', 'Geschichte');
      expect(component.history).toEqual(testWikiGermanServiceResponse.history);
      expect(component.title).toEqual(testWikiGermanServiceResponse.title);
      expect(component.langlinks).toEqual(testWikiGermanServiceResponse.langlinks);
    });
```

File: capstone-app/src/app/modules/history/wiki-media/wiki-media.component.ts [hunk 21]
```diff
@@ -18,7 +22,15 @@ export class WikiMediaComponent implements OnChanges, OnInit {
   loading = true;
   body: string;
   urls = new Map();
-  cityName: string;
+  langlinks = new Map();
+  languages: any = Languages as any;
```
Comment (mbehrlich) [inline capstone-app/src/app/modules/history/wiki-media/wiki-media.component.ts]: Not crazy about use of "any", why is this necessary? If absolutely necessary, I think unknown type would be preferred.

File: capstone-app/src/app/modules/history/wiki-media/wiki-media.component.ts [hunk 22]
```diff
@@ -42,21 +53,71 @@ export class WikiMediaComponent implements OnChanges, OnInit {
...
   }
+
+  changeLanguage(queryString: string, languagePrefix: string, wordForHistory: string) {
+    this.wikiService.searchNewLang(queryString, languagePrefix, wordForHistory)
+      .subscribe((result: WikiServiceResult) => {
+        this.history = result.history;
+        console.log(this.history);
+        if (!(this.history) || (this.history.includes("CPU time usage:"))){
+          this.loading = true;
+          this.error = "Unfortunately, the language you requested does not have an available tranlation for this page. Please select a different language.";
```
Comment (mbehrlich) [inline capstone-app/src/app/modules/history/wiki-media/wiki-media.component.ts]: Use single quotes. Also, break this up into two lines, as it's way to long. You can do this with +
```
this.error = 'Unfortunately, the language ' +
    'you requested does not have...';
```

File: capstone-app/src/app/modules/history/wiki-results.service.ts [hunk 23]
```diff
@@ -16,27 +16,35 @@ export class WikiResultsService {
...
         ),
         map(history => {
           if (history.parse.text) {
-            const historyResult = this.fixString(history.parse.text['*']);
+            const langlinks = new Map<string, Map<string, string>>();
+            for (let link of Object.values(history.parse.langlinks)) {
+                let links = new Map<string, string>();
+                links.set('langName', link.langname);
+                links.set('url', link.url);
+                links.set('searchQuery', link["*"]);
```
Comment (mbehrlich) [inline capstone-app/src/app/modules/history/wiki-results.service.ts:44]: Use single quotes

File: capstone-app/src/app/modules/history/wiki-results.service.ts [hunk 24]
```diff
@@ -46,11 +54,40 @@ export class WikiResultsService {
...
+    return this.http.get<WikiSearchResult>(
+            `${WikiResultsService.URL_BEGINNING}${language}${WikiResultsService.URL_MIDDLE}${query}${WikiResultsService.URL_END}`
+          ).pipe(map(history => {
+          if (history.parse.text) {
+            const langlinks = new Map<string, Map<string, string>>();
+            for (const link of Object.values(history.parse.langlinks)) {
+                const links = new Map<string, string>();
+                links.set('langName', link.langname);
+                links.set('url', link.url);
+                links.set('searchQuery', link["*"]);
```
Comment (mbehrlich) [inline capstone-app/src/app/modules/history/wiki-results.service.ts]: Single quotes

File: capstone-app/src/app/modules/history/wiki-results.service.ts [hunk 25]
```diff
@@ -46,11 +54,40 @@ export class WikiResultsService {
...
+        }),
+        catchError(() => {
+          console.error('API did not return a valid response.');
+          return of({title: '', history: ''});
+        })
+      );
+  }
+
+
+  fixString(text: string, wordForHistory: string, language: string): {history: string, furtherReading?: Map<string, string>} {
```
Comment (mbehrlich) [inline capstone-app/src/app/modules/history/wiki-results.service.ts:86]: Very confused about "furtherReading" optional parameter. Where is it used? Why is it redefined below?

Comment (mbehrlich) [inline capstone-app/src/app/modules/history/wiki-results.service.ts]: Recommend defining this interface above and referencing it here:

```
interface HistoryObjectThing {
  history: string;
  furtherReading?: Map<string, string>;
}
```

File: capstone-app/src/app/modules/history/wiki-results.service.ts [hunk 26]
```diff
@@ -78,17 +115,18 @@ export class WikiResultsService {
...
+  findHrefs(text: string, languageAbreviation: string): Map<string, string> {
     const urls = new Map<string, string>();
     const el = document.createElement('p');
     el.innerHTML = text;
     const hrefs = Array.from(el.querySelectorAll('a'));
     let count = 1;
     for (const h of hrefs) {
-      if (count <= 12 && !(h.getAttribute('href').charAt(0) === '#' || (h.getAttribute('title') && h.getAttribute('title').startsWith('Edit')))) {
+      if (count <= 12 && !(h.getAttribute('href').charAt(0) === '#' ||
+      (h.getAttribute('title') && h.getAttribute('title').startsWith('Edit')))) {
```
Comment (mbehrlich) [inline capstone-app/src/app/modules/history/wiki-results.service.ts]: indent

**Review by mbehrlich — APPROVED** [review]

File: capstone-app/src/app/modules/history/wiki-media/wiki-media.component.spec.ts (hunk 20 above)
Comment (Sarahhud) [reply capstone-app/src/app/modules/history/wiki-media/wiki-media.component.spec.ts:145]: Oh, I see. I would prefer to leave them separate so if they fail it's more clear which one is failing. Is that okay?

**Review by smontana30 — APPROVED** [review]

File: capstone-app/src/app/modules/history/wiki-media/wiki-media.component.ts (hunk 21 above)
Comment (Sarahhud) [reply capstone-app/src/app/modules/history/wiki-media/wiki-media.component.ts]: Using un-kown works but if I try casting it to anything else it doesn't seem to run properly.

Comment (mbehrlich) [reply capstone-app/src/app/modules/history/wiki-media/wiki-media.component.ts]: Why is it necessary in the first place?

Comment (mbehrlich) [reply capstone-app/src/app/modules/history/wiki-media/wiki-media.component.ts]: Nevermind, looks like you fixed it.

File: capstone-app/src/app/modules/history/wiki-media/wiki-media.component.ts [hunk 27]
```diff
@@ -18,7 +22,14 @@ export class WikiMediaComponent implements OnChanges, OnInit {
   loading = true;
   body: string;
   urls = new Map();
-  cityName: string;
+  langlinks = new Map();
+  languages: {[language: string]: string} = Languages as {[language: string]: string};
```
Comment (NicholasCobbFromGoogle) [inline capstone-app/src/app/modules/history/wiki-media/wiki-media.component.ts]: Let's also make an interface for this type:
export interface LanguageData {
  [language: string]: string;
}

File: capstone-app/src/app/modules/history/wiki-media/wiki-media.component.ts [hunk 28]
```diff
@@ -42,21 +52,77 @@ export class WikiMediaComponent implements OnChanges, OnInit {
...
         }
     });
   }
+
+  changeLanguage(queryString: string, languagePrefix: string, wordForHistory: string) {
+    this.wikiService.searchNewLang(queryString, languagePrefix, wordForHistory)
+      .subscribe((result: WikiServiceResult) => {
+        this.history = result.history;
+        console.log(this.history);
+        if (!(this.history)){
```
Comment (NicholasCobbFromGoogle) [inline capstone-app/src/app/modules/history/wiki-media/wiki-media.component.ts]: no need for the parens around this.history

File: capstone-app/src/app/modules/history/wiki-media/wiki-media.component.ts [hunk 29]
```diff
@@ -42,21 +52,77 @@ export class WikiMediaComponent implements OnChanges, OnInit {
...
           console.error('API did not return a valid response.');
         }
     });
   }
+
+  changeLanguage(queryString: string, languagePrefix: string, wordForHistory: string) {
+    this.wikiService.searchNewLang(queryString, languagePrefix, wordForHistory)
+      .subscribe((result: WikiServiceResult) => {
+        this.history = result.history;
+        console.log(this.history);
```
Comment (NicholasCobbFromGoogle) [inline capstone-app/src/app/modules/history/wiki-media/wiki-media.component.ts]: let's get rid of the console.log

File: capstone-app/src/app/modules/history/wiki-media/wiki-media.component.ts [hunk 30]
```diff
@@ -42,21 +52,77 @@ export class WikiMediaComponent implements OnChanges, OnInit {
...
+          + 'The data being shown is what the Wikipedia page returned.';
+        }
+    });
+  }
+
+  onChangeLanguage(prefix: string) {
+    if (prefix == this.prevPreFix) {
+      this.changeLanguage(this.prevQuery, this.prevPreFix, this.prevWordForHistory);
+    }
+    else if (prefix == 'en' && !(this.langlinks.get('en'))) {
```
Comment (NicholasCobbFromGoogle) [inline capstone-app/src/app/modules/history/wiki-media/wiki-media.component.ts]: nit: else if (prefix === 'en' && !this.langlinks.get('en')) {

File: capstone-app/src/app/modules/history/wiki-media/wiki-media.component.ts [hunk 31]
```diff
@@ -42,21 +52,77 @@ export class WikiMediaComponent implements OnChanges, OnInit {
...
+        if (this.history.includes("CPU time usage:")){
+          this.loading = false;
+          this.error = 'Unfortunately, the language you requested has not been fully translated. '
+          + 'The data being shown is what the Wikipedia page returned.';
+        }
+    });
+  }
+
+  onChangeLanguage(prefix: string) {
+    if (prefix == this.prevPreFix) {
```
Comment (NicholasCobbFromGoogle) [inline capstone-app/src/app/modules/history/wiki-media/wiki-media.component.ts]: nit: we should never use == in javascript/typescript, always ===.

== is unintuitive due to the loose typing of javascript and the gymnastics around the language attempting to cast values while checking equality.
https://dorey.github.io/JavaScript-Equality-Table/

File: capstone-app/src/app/modules/history/wiki-media/wiki-media.component.ts [hunk 32]
```diff
@@ -42,21 +52,77 @@ export class WikiMediaComponent implements OnChanges, OnInit {
...
+      const language = this.langlinks.get(prefix);
+      this.query = language.get('searchQuery');
+      this.prevQuery = this.query;
+      this.prevWordForHistory = 'History';
+      this.changeLanguage(this.query, 'en', 'History');
+    }
+    else {
+      const language = this.langlinks.get(prefix);
+      console.log(this.langlinks);
+      console.log(prefix);
```
Comment (NicholasCobbFromGoogle) [inline capstone-app/src/app/modules/history/wiki-media/wiki-media.component.ts]: nit: remove console.logs

File: capstone-app/src/app/modules/history/wiki-results.service.spec.ts [hunk 33]
```diff
@@ -4,6 +4,7 @@ import { WikiResultsService } from './wiki-results.service';
 import { WikiSearchResult } from './WikiSearchTemplate';
 import MockWikiResponse from 'testing/mock-wiki-response.json';
 
+
```
Comment (NicholasCobbFromGoogle) [inline capstone-app/src/app/modules/history/wiki-results.service.spec.ts]: nit: remove extra newline

File: capstone-app/src/app/modules/history/wiki-results.service.ts [hunk 34]
```diff
@@ -6,37 +6,48 @@ import { WikiServiceResult } from './WikiServiceResult';
 import { HttpClient } from '@angular/common/http';
 import { Observable, of } from 'rxjs';
 import MockWikiResponse from 'testing/mock-wiki-response.json';
-
+interface HistoryFixString{
+  history: string;
+  furtherReading?: Map<string, string>;
+}
```
Comment (NicholasCobbFromGoogle) [inline capstone-app/src/app/modules/history/wiki-results.service.ts:13]: nit: let's add a newline each before and after this interface definition

File: capstone-app/src/app/modules/history/wiki-results.service.ts [hunk 35]
```diff
@@ -46,49 +57,80 @@ export class WikiResultsService {
...
+        }),
+        catchError(() => {
+          console.error('API did not return a valid response.');
+          return of({title: '', history: ''});
+        })
+      );
+  }
+
+
+  fixString(text: string, wordForHistory: string, language: string): HistoryFixString {
```
Comment (NicholasCobbFromGoogle) [inline capstone-app/src/app/modules/history/wiki-results.service.ts:91]: let's remove the console.log calls in this method. (I am indifferent to console.error, however)

File: capstone-app/src/app/modules/material/material.module.ts [hunk 36]
```diff
@@ -13,6 +15,8 @@ const material = [
   MatInputModule,
   MatMenuModule,
   MatSelectModule,
+  MatCardModule,
+  MatNativeDateModule
```
Comment (NicholasCobbFromGoogle) [inline capstone-app/src/app/modules/material/material.module.ts]: nit: let's add a comma at the end of the line

**Review by NicholasCobbFromGoogle — APPROVED** [review]

File: capstone-app/src/app/modules/history/wiki-media/wiki-media.component.ts (hunk 32 above)
Comment (Sarahhud) [reply capstone-app/src/app/modules/history/wiki-media/wiki-media.component.ts]: Sorry about all of those, I totally forgot to remove them on my last commit.

**Review by billyporter — APPROVED** [review]
